services/motion/explore/explore.go

```diff
@@ -0,0 +1,576 @@
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return true, nil
+			}
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return true, nil
+			}
+		}
+	}
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorObject,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	obstaclePollingFrequencyHz float64,
+) {
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPose := spatialmath.NewZeroPose()
+
+			// Look for new transient obstacles and add to worldState
+			worldState, err := ms.generateTransientWorldState(ctx, obstacleDetectors)
+			if err != nil {
+				ms.logger.Debugf("issue occurred generating transient worldState: %v", err)
+			}
+
+			// Check motionplan plan for transient obstacles
+			collisionPose, err := motionplan.CheckPlan(
+				kb.Kinematics(),
+				plan,
+				worldState,
+				ms.frameSystem,
+				currentPose,
+				make([]referenceframe.Input, len(kb.Kinematics().DoF())),
+				spatialmath.NewZeroPose(),
+				ms.logger,
+			)
+			if err != nil {
+				// If an obstacle is detected, check if its within the valid obstacle distance to trigger an
+				// end to the checkForObstacle loop
+				if collisionPose.Point().Distance(currentPose.Point()) < validObstacleDistanceMM {
+					ms.logger.Debug("collision found")
+					ms.obstacleResponseChan <- moveResponse{success: true}
+					return
+				}
+				ms.logger.Debug("collision found but outside of range")
+			}
+		}
+	}
+}
+
+// executePlan will carry out the desired motionplan plan.
+func (ms *explore) executePlan(ctx context.Context, kb kinematicbase.KinematicBase, plan motionplan.Plan) {
+	// Iterate through motionplan plan
+	for i := 1; i < len(plan); i++ {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if inputEnabledKb, ok := kb.(inputEnabledActuator); ok {
```

Comment:
If not, throw an error.

Reply:
Added error being sent to channel: 			

`ms.executionResponseChan <- moveResponse{err: errors.New("unable to case kinematic base to inputEnabledActuator")}`